VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,850 @@
+must be called for every phase method that is being overridden.
+8.  Transaction classes are usually logical representations of the data
+communicated over one or more interfaces executing a particular protocol.
+These classes:
+    *   Should extend `uvm_sequence_item`.
+    *   Should contain only information that is relevant to the data being
+    transmitted over an interface, and not to the means of transmit.
+    *   Should result in a legal transaction when randomized.
+
+
+### UVM `new()` functions
+
+The `new()` function of any class derived from `uvm_object` must have the
+default value of its `name` argument set to `""`, since the default name
+argument will be set by `type_id::create()`.
+The `new()` function of any class not derived from `uvm_object` must not
+have default values set for its arguments.
+Examples:
+
+:+1:
+```systemverilog
+// class derived from uvm_object with correct default value
+function new (string name = "");
+  super.new(name);
+endfunction : new
+
+// class derived from uvm_component with no default values
+function new(string name, uvm_component parent);
+  super.new(name, parent);
+endfunction : new
+```
+
+:-1:
+```systemverilog
+// Incorrect default value used for class derived from uvm_object
+function new(string name = "my_rand_seq");
+  super.new(name);
+endfunction : new
+
+// Incorrect because default values for class derived from uvm_component are provided
+function new(string name = "", uvm_component parent);
+  super.new(name, parent);
+endfunction : new
+```
+
+
+### `uvm_scoreboard` usage
+
+Use `uvm_tlm_analysis_fifo` - can be directly connected to analysis ports,
+and has an unbounded size so that writes will always succeed without blocking,
+in situations where transactions need to be stored for some time before being
+consumed by the scoreboard for processing.
+
+
+### `uvm_agent` usage
+
+1. Interface agents should contain only a sequencer, driver, monitor, and config object.
+2. Use one agent for every interface of the DUT, with an optional sequencer and
+driver whose instantiation is determined by the value of `<agent_cfg>.is_active`.
+This is placed in the agent config object to allow for better access throughout the testbench.
+3. For any DUT interfaces that have multiple operation modes
+(such as host and device modes), two drivers should be created, one for each mode,
+and arbitration done in the `uvm_agent::build_phase` function to determine
+which driver to instantiate.
+Each driver should have different sequences that correspond to it.
+
+
+### `uvm_driver` usage
+
+1. A driver should only communicate with one sequencer on one SystemVerilog
+interface, and should not have any analysis ports.
+2. A driver should not randomize any data any data received from transaction
+items sent through analysis ports.
+3. A driver must assign `X` to any buses it controls during "don't-care" clock
+cycles in which no valid transaction is present.
+
```

Comment:
Should we add a note here that contrasts this with what the RTL guide recommends in terms of usage of `X`?

Reply:
I feel that may not be needed, since the RTL guide explictly states that the `X` guidelines should apply to RTL designs only, whereas here we only refer to code that lives within the UVM framework. What do you think?

Reply:
That is true, it explicitly states that the guidance applies to RTL. Ok if we omit it here.